lib/burnside/convert-expr.cc

```diff
@@ -412,6 +426,11 @@ class ExprLowering {
     case Ev::LogicalOperator::Neqv:
       result = createCompareOp<M::CmpIOp>(op, M::CmpIPredicate::ne);
       break;
+    case Ev::LogicalOperator::Not:
+      // LogicalOperations are binary operations. Expr for Not is
+      // evaluate::Not<KIND>.
```

Comment:
I'm confused. If `Ev::LogicalOperator`s are binary operations, why is `Not` here?  Also, if it can appear in a `LogicalOperation<KIND>`, why is it unhandled here?

Reply:
Should this call `genval(const Ev::Not<KIND> &op)` in case a parse tree actually has this pattern?

Reply:
I think your confusion comes from type names that look like each other: `evaluate::LogicalOperation` are binary operations, not `common::LogicalOperator`.

`.NOT.` was added to `common::LogicalOperator` in:
https://github.com/flang-compiler/f18/pull/815/files#diff-038ccf1943dfbabce4226c2833f6ef6bR44

This addition intention is not to change `lib/evaluate` expression structure, it is motivated by parsing/semantics changes around logical operators (mainly to handle the different abbreviations that exist for logical operators, see https://github.com/flang-compiler/f18/pull/815 for more details).

Because  `evaluate::LogicalOperation` happens to [contain a `common::LogicalOperator`](https://github.com/flang-compiler/f18/blob/4a6d6f184bcf54de971f30df0bf6649e2768016f/lib/evaluate/expression.h#L406), this change had the side effect of opening the door of `.NOT.` being encoded in `evaluate::LogicalOperation`, but this is considered an internal error because `evaluate::LogicalOperation` encodes binary operations. An assert was actually inserted in the front-end during the change to enforce this:
https://github.com/flang-compiler/f18/pull/815/files#diff-f6631e851cd4491a4297d7e3d2f884ebR2462

Hence, I think we also want to have the same assert in the bridge: the parse-tree cannot have this pattern.

Reply:
I think your confusion comes from type names that look like each other: `evaluate::LogicalOperation` are binary operations, not `common::LogicalOperator`.

`.NOT.` was added to `common::LogicalOperator` in:
https://github.com/flang-compiler/f18/pull/815/files#diff-038ccf1943dfbabce4226c2833f6ef6bR44

This addition intention is not to change `lib/evaluate` expression structure, it is motivated by parsing/semantics changes around logical operators (mainly to handle the different abbreviations that exist for logical operators, see https://github.com/flang-compiler/f18/pull/815 for more details).

Because  `evaluate::LogicalOperation` happens to [contain a `common::LogicalOperator`](https://github.com/flang-compiler/f18/blob/4a6d6f184bcf54de971f30df0bf6649e2768016f/lib/evaluate/expression.h#L406), this change had the side effect of opening the door of `.NOT.` being encoded in `evaluate::LogicalOperation`, but this is considered an internal error because `evaluate::LogicalOperation` encodes binary operations. An assert was actually inserted in the front-end during the change to enforce this:
https://github.com/flang-compiler/f18/pull/815/files#diff-f6631e851cd4491a4297d7e3d2f884ebR2462

Hence, I think we also want to have the same assert in the bridge: the parse-tree cannot have this pattern.

Reply:
Thanks for the explanation. It's an odd exception, but the remedy sounds correct.